File: backend/app/utils/pricing.py
"""Effective (sale) price helpers — the single source of truth for what a product
actually costs. Every charge and quote must route through ``effective_price`` so a
manually-entered sale price is honoured consistently (order, cart, re-order, Qiki).
"""

from decimal import ROUND_HALF_UP, Decimal


def effective_price(price: Decimal, sale_price: Decimal | None) -> Decimal:
    """Return the price actually charged.

    The sale price wins only when it is a valid discount (``0 < sale_price < price``);
    otherwise the regular list price applies.
    """
    if sale_price is not None and Decimal(0) < sale_price < price:
        return sale_price
    return price


def discount_percent(price: Decimal, sale_price: Decimal | None) -> int | None:
    """Return the integer discount percentage when a valid sale price exists, else None."""
    if sale_price is None or not (Decimal(0) < sale_price < price):
        return None
    return int(((price - sale_price) / price * 100).quantize(Decimal(1), rounding=ROUND_HALF_UP))

File: frontend/lib/utils/pricing.ts
import type { Product } from '@/types/product'

/** The price actually charged: the sale price when it is a valid discount below list. */
export function effectivePriceValue(price: number, salePrice: number | null): number {
  return salePrice != null && salePrice > 0 && salePrice < price ? salePrice : price
}

/** Integer discount percentage when a valid sale price exists, else null. */
export function discountPercentValue(price: number, salePrice: number | null): number | null {
  if (salePrice == null || !(salePrice > 0 && salePrice < price)) return null
  return Math.round(((price - salePrice) / price) * 100)
}

type PricedProduct = Pick<Product, 'price' | 'sale_price'>

function toNumber(value: string | number | null): number | null {
  if (value == null) return null
  const parsed = typeof value === 'string' ? Number.parseFloat(value) : value
  return Number.isNaN(parsed) ? null : parsed
}

/** Effective price for a product (list price string + optional sale price string). */
export function effectivePrice(product: PricedProduct): number {
  return effectivePriceValue(Number(product.price), toNumber(product.sale_price))
}

/** Discount percentage for a product, or null when it is not on sale. */
export function discountPercent(product: PricedProduct): number | null {
  return discountPercentValue(Number(product.price), toNumber(product.sale_price))
}
